src/ParallelAlgorithms/NonlinearSolver/NewtonRaphson/ElementActions.hpp

```diff
@@ -0,0 +1,508 @@
+          array_index,
+          db::get<Convergence::Tags::IterationId<OptionsGroup>>(box),
+          db::get<NonlinearSolver::Tags::StepLength<OptionsGroup>>(box));
+    }
+
+    // Apply the correction that the linear solve has determined to attempt
+    // improving the nonlinear solution
+    db::mutate<fields_tag>(
+        make_not_null(&box),
+        [](const auto fields, const auto& correction, const double step_length,
+           const auto& globalization_fields) {
+          *fields = globalization_fields + step_length * correction;
+        },
+        db::get<correction_tag>(box),
+        db::get<NonlinearSolver::Tags::StepLength<OptionsGroup>>(box),
+        db::get<globalization_fields_tag>(box));
+
+    return {std::move(box)};
+  }
+};
+
+// Between `PerformStep` and this action the nonlinear operator has been applied
+// to the updated fields. The residual is up-to-date because it is a compute
+// tag, so at this point we need to check if the step has sufficiently reduced
+// the residual. We perform a global reduction to the `ResidualMonitor` for this
+// purpose.
+template <typename FieldsTag, typename OptionsGroup, typename Label>
+struct ContributeToResidualMagnitudeReduction {
+ private:
+  using fields_tag = FieldsTag;
+  using nonlinear_residual_tag =
+      db::add_tag_prefix<NonlinearSolver::Tags::Residual, fields_tag>;
+
+ public:
+  template <typename DbTagsList, typename... InboxTags, typename Metavariables,
+            typename ArrayIndex, typename ActionList,
+            typename ParallelComponent>
+  static std::tuple<db::DataBox<DbTagsList>&&> apply(
+      db::DataBox<DbTagsList>& box,
+      const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+      Parallel::GlobalCache<Metavariables>& cache,
+      const ArrayIndex& array_index, const ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) noexcept {
+    const auto& residual = db::get<nonlinear_residual_tag>(box);
+    const double local_residual_magnitude_square =
+        LinearSolver::inner_product(residual, residual);
+    ResidualReductionData reduction_data{
+        db::get<Convergence::Tags::IterationId<OptionsGroup>>(box),
+        db::get<NonlinearSolver::Tags::Globalization<
+            Convergence::Tags::IterationId<OptionsGroup>>>(box),
+        local_residual_magnitude_square,
+        db::get<NonlinearSolver::Tags::StepLength<OptionsGroup>>(box)};
+    Parallel::contribute_to_reduction<
+        CheckResidualMagnitude<FieldsTag, OptionsGroup, ParallelComponent>>(
+        std::move(reduction_data),
+        Parallel::get_parallel_component<ParallelComponent>(cache)[array_index],
+        Parallel::get_parallel_component<
+            ResidualMonitor<Metavariables, FieldsTag, OptionsGroup>>(cache));
+    return {std::move(box)};
+  }
+};
+
+// Wait for the `ResidualMonitor` to broadcast if it has determined the step
+// was sufficient or not. If it was, we just proceed to `CompleteStep`. If it
```

Comment:
again, sufficient seems awkward. Maybe `acceptable`? Or `the step decreased the residual sufficiently`, which I think is clearest